evaluation/evaluate.py

```diff
@@ -0,0 +1,273 @@
+import concurrent.futures
+import json
+import logging
+import os
+import time
+from concurrent.futures import ThreadPoolExecutor
+from pathlib import Path
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import requests
+from promptflow.core import AzureOpenAIModelConfiguration
+from rich.progress import track
+
+from evaluation import service_setup
+from evaluation.evaluate_metrics import metrics_by_name
+from evaluation.utils import load_jsonl
+
+EVALUATION_RESULTS_DIR = "gpt_evaluation"
+
+logger = logging.getLogger("evaluation")
+
+
+def send_question_to_target(question: str, url: str, parameters: dict = {}, raise_error=True) -> dict:
+    """Send a question to the ask endpoint and return the response."""
+    headers = {
+        "Content-Type": "application/json",
+    }
+    body = {
+        "messages": [{"content": question, "role": "user"}],
+        "context": parameters,
+    }
+
+    try:
+        r = requests.post(url, headers=headers, json=body)
+        r.encoding = "utf-8"
+        latency = r.elapsed.total_seconds()
+
+        if not r.ok:
+            raise ConnectionError(f"Request to target {url} failed with status code {r.status_code}: {r.reason} \n")
```

Comment:
@GrayNekoBean what you're trying to do there already exists in the library. `response.raise_for_status()` will raise the error that caused the request to be unsuccessful. In fact, `ok` is [using it under the hood](https://github.com/psf/requests/blob/0e322af87745eff34caffe4df68456ebc20d9068/src/requests/models.py#L755-L767). What's more though, this is not a `ConnectionError`, it is specifically an `HTTPError`. It's already there, why would you raise a different error than the one the library returned?

Reply:
Thx, I still have a lot to learn.